spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,492 @@
+# This would shift the training distribution to over-represent certain minority populations.
+# Intuitively, we'd like to show the model more `short_link` examples so that the representation is better suited to handle them.
+#
+# A technique like upsampling might work with a small number of slices, but with hundreds or thousands or production slices, it could quickly become intractable to tune upsampling weights per slice.
+# In the following section, we show a modeling approach that we call _Slice-based Learning,_ which handles numerous slices using with slice-specific representation learning.
+
+# %% [markdown]
+# ### Write additional slicing functions (SFs)
+#
+# We'll take inspiration from the labeling tutorial to write additional slicing functions.
+
+# %%
+from snorkel.slicing import SlicingFunction, slicing_function, nlp_slicing_function
+from snorkel.preprocess import preprocessor
+
+
+# Keyword-based SFs
+def keyword_lookup(x, keywords):
+    return any(word in x.text.lower() for word in keywords)
+
+
+def make_keyword_sf(keywords):
+    return SlicingFunction(
+        name=f"keyword_{keywords[0]}",
+        f=keyword_lookup,
+        resources=dict(keywords=keywords),
+    )
+
+
+keyword_subscribe = make_keyword_sf(keywords=["subscribe"])
+keyword_please = make_keyword_sf(keywords=["please", "plz"])
+
+
+# Leverage @nlp_slicing_function
+@nlp_slicing_function()
+def has_person_nlp(x):
+    """Ham comments mention specific people and are short."""
+    return len(x.doc) < 20 and any([ent.label_ == "PERSON" for ent in x.doc.ents])
+
+
+# Regex-based SF
+@slicing_function()
+def regex_check_out(x):
+    return bool(re.search(r"check.*out", x.text, flags=re.I))
+
+
+# Heuristic-based SF
+@slicing_function()
+def short_comment(x):
+    """Ham comments are often short, such as 'cool video!'"""
+    return len(x.text.split()) < 5
+
+
+# Leverage preprocessor in SF
+from textblob import TextBlob
+
+
+@preprocessor(memoize=True)
+def textblob_sentiment(x):
+    scores = TextBlob(x.text)
+    x.polarity = scores.sentiment.polarity
+    return x
+
+
+@slicing_function(pre=[textblob_sentiment])
+def textblob_polarity(x):
+    return x.polarity > 0.9
+
+
+extra_sfs = [
+    keyword_subscribe,
+    keyword_please,
+    regex_check_out,
+    short_comment,
+    has_person_nlp,
+    textblob_polarity,
+]
+
+sfs = [short_link] + extra_sfs
+
+# %%
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %%
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
+add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
+add_slice_labels(dl_test, spam_task, S_test, slice_names)
+
+# %% [markdown]
+# Like we saw above, we'd like to visualize examples in the slice.
+
+# %%
+polarity_df = slice_dataframe(df_valid, textblob_polarity)
```

Comment:
why are the labels zero now? worth a comment

Reply:
i.e. feels like it might mean "slice_label", just clarify

Reply:
clarified an explanation for 0 labels in relation to high-polarity examples